File: bongf/python/level2/truck_crossing_the_bridge.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    trucks = deque(truck_weights)
    bridge = deque([0] * bridge_length)
    weight_bridge = 0
    time = 0

    while True :
        weight_bridge -= bridge.popleft()
        bridge.append(0)
        if(trucks != deque([]) and weight_bridge + trucks[0] <= weight) :
            new = trucks.popleft()
            bridge.pop()
            bridge.append(new)
            weight_bridge += new 

        time += 1
        ## 조건문 순서 바꿔주면 fail 된다. [0] * bridege_length를 확인하는 부분이 오래 걸리나봄 
        if(trucks == deque([]) and bridge == deque([0] * bridge_length)) :
            break
    return time

## while 문을 truck이 비었을 때로 설정하고 마지막에 그 브릿지만큼 더해준 것이 똑띠, reverse 써서 queue 안만들어도 된 것 똑띠 
def solution2(bridge_length, weight, truck_weights):
    bridge = deque(0 for _ in range(bridge_length))
    total_weight = 0
    step = 0
    truck_weights.reverse()

    while truck_weights:
        total_weight -= bridge.popleft()
        if total_weight + truck_weights[-1] > weight:
            bridge.append(0)
        else:
            truck = truck_weights.pop()
            bridge.append(truck)
            total_weight += truck
        step += 1

    step += bridge_length

    return step


print(solution(2, 10, [7,4,5,6]))
print(solution(100, 100, [10]))
print(solution(100, 100, [10,10,10,10,10,10,10,10,10,10]))

File: programmers/java/src/level2/TruckCrossingTheBridge.java
package level2;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.stream.Collectors;

public class TruckCrossingTheBridge {
    public static void main(String[] args) {
        System.out.println(solution(2, 10, new int[]{7,4,5,6}));
    }

    public static int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> trucks = Arrays.stream(truck_weights).boxed().collect(Collectors.toCollection(LinkedList::new));
        Integer[] temp = new Integer[bridge_length];
        Arrays.fill(temp, 0);
        Queue<Integer> bridge = new LinkedList<>(Arrays.asList(temp));
        int time = 0;
        int weightOnTheBridge = 0;

        while (!trucks.isEmpty()) {
            weightOnTheBridge -= bridge.poll();
            if (weightOnTheBridge + trucks.peek() <= weight) {
                int truck = trucks.poll();
                bridge.offer(truck);
                weightOnTheBridge += truck;
            } else {
                bridge.offer(0);
            }
            time++;
        }

        time += bridge_length;
        return time;
    }

    // for문 돌면서 que에 elements 넣는 것이 더 간단해 보임
    //클래스로 구현한 친구
    public int solution2(int bridgeLength, int weight, int[] truckWeights) {
        Queue<Truck> waitQ = new LinkedList<>();
        Queue<Truck> moveQ = new LinkedList<>();

        for (int t : truckWeights) {
            waitQ.offer(new Truck(t));
        }

        int answer = 0;
        int curWeight = 0;

        while (!waitQ.isEmpty() || !moveQ.isEmpty()) {
            answer++;

            if (moveQ.isEmpty()) {
                Truck t = waitQ.poll();
                curWeight += t.weight;
                moveQ.offer(t);
                continue;
            }

            for (Truck t : moveQ) {
                t.moving();
            }

            if (moveQ.peek().move > bridgeLength) {
                Truck t = moveQ.poll();
                curWeight -= t.weight;
            }

            if (!waitQ.isEmpty() && curWeight + waitQ.peek().weight <= weight) {
                Truck t = waitQ.poll();
                curWeight += t.weight;
                moveQ.offer(t);
            }
        }

        return answer;
    }

    class Truck {
        int weight;
        int move;

        public Truck(int weight) {
            this.weight = weight;
            this.move = 1;
        }

        public void moving() {
            move++;
        }
    }
}
